component/loki/source/file/file.go

```diff
@@ -0,0 +1,277 @@
+		return nil, err
+	}
+
+	c := &Component{
+		opts:    o,
+		metrics: newMetrics(o.Registerer),
+
+		handler:   make(chan api.Entry),
+		receivers: args.ForwardTo,
+		posFile:   positionsFile,
+		readers:   make(map[string]Reader),
+	}
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+// TODO(@tpaschalis). Should we periodically re-check? What happens if a target
+// comes alive _after_ it's been passed to us and we never receive another
+// Update()? Or should it be a responsibility of the discovery component?
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		level.Info(c.opts.Logger).Log("msg", "loki.source.file component shutting down, stopping readers")
+		for _, r := range c.readers {
+			r.Stop()
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			for _, receiver := range c.receivers {
+				receiver <- entry
+			}
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.args = newArgs
+	c.receivers = newArgs.ForwardTo
+
+	oldPaths := make(map[string]struct{})
+
+	// Stop all readers and recreate them below. This avoids the issue we saw
+	// with stranded wrapped handlers staying behind until they were GC'ed and
+	// sending duplicate message to the global handler. It also makes sure that
+	// we update everything with the new labels. Simply zeroing out the
+	// c.readers map did not work correctly to shut down the wrapped handlers
+	// in time.
+	// TODO (@tpaschalis) We should be able to optimize this somehow and eg.
+	// cache readers for paths we already know about, and whose labels have not
+	// changed. Once we do that we should:
+	// a) Call to c.pruneStoppedReaders to give cached but errored readers a
+	// chance to restart.
+	// b) Stop tailing any files that were no longer in the new targets
+	// and conditionally remove their readers only by calling toStopTailing
+	// and c.stopTailingAndRemovePosition.
+	for p, r := range c.readers {
+		oldPaths[p] = struct{}{}
+		r.Stop()
+	}
+	c.readers = make(map[string]Reader)
+
+	if len(newArgs.Targets) == 0 {
+		level.Debug(c.opts.Logger).Log("msg", "no files targets were passed, nothing will be tailed")
+		return nil
+	}
+
+	for _, target := range newArgs.Targets {
+		path := target[pathLabel]
+		c.reportSize(path)
+
+		var labels = make(model.LabelSet)
+		for k, v := range target {
+			if strings.HasPrefix(k, "__") {
+				continue
+			}
+			labels[model.LabelName(k)] = model.LabelValue(v)
+		}
+
+		handler := api.AddLabelsMiddleware(labels).Wrap(api.NewEntryHandler(c.handler, func() {}))
+
+		reader, err := c.startTailing(path, handler)
+		if err != nil {
+			continue // TODO (@tpaschalis) return err maybe?
+		}
```

Comment:
Is this the part where you're worried we might want to re-sync targets on an interval? I think it's ok to retry tailing failed targets after a certain period. 

I definitely don't think we should return the error here; we don't want one failed target to cause the sync to fail and disrupt log collection for other files. 

Also: we should make sure the error is _at least_ being logged here, but we'll also want to expose these targets as debug information just like we do for prometheus.scrape so people can see tailing errors.

Reply:
Good points both about logging and exposing debug information about these targets, I'll take care of it.

Yes, that's the point. For example here, if we receive a target for a __path__ that doesn't yet exist, then we will never know about it until a new Update comes in.

cc @mattdurham  Some of Promtail's SD mechanisms use a `refresh_interval` that re-read the target set. I can see an argument that this should be the responsibility of a discovery component to provide a 'correct' list of targets and update accordingly, but I'd personally lean to adding a relatively longer `refresh_interval` ticker in the Run's select that calls Update every now and then. WDYT?

Reply:
This sounds similar to local.file and remote.s3 components that refresh and then update. Maybe steal from those?